### PR TITLE
Adding Mobile UX for Asset Bundle downloads

### PR DESCRIPTION
**Pull request by mrbobbybobberson**
- Broke the asset bundle send/receive into chunks so progress can be communicated in UX and the main UI threads don't hang during the whole send.
- Gave asset bundles identities (versions) to pave the way for bundle caching and to allow for a "friendly name" to be displayed in the UX.

### CODE REVIEW

File: samples/Build2019Demo.Unity/Assets/Generated.StateSynchronization.AssetCaches/Resources/Android/spectatorview.manifest.bytes [hunk 1]
```diff
@@ -52,7 +52,6 @@ Assets:
 - Assets/Generated.StateSynchronization.AssetCaches/Resources/MaterialPropertyAssetCache.asset
 - Assets/Generated.StateSynchronization.AssetCaches/Resources/AudioMixerGroupAssetCache.asset
 - Assets/Generated.StateSynchronization.AssetCaches/Resources/TextureAssetCache.asset
-- Assets/Demo/PlaygroundButtons/Shaders/Rounded_Edge_Thick.shader
```
Comment (mrbobbybobberson) [inline samples/Build2019Demo.Unity/Assets/Generated.StateSynchronization.AssetCaches/Resources/Android/spectatorview.manifest.bytes]: @matthejo , any idea why when you generate bundles, you include this, but when I generate them, I exclude it?

File: src/SpectatorView.Unity/Assets/SpectatorView/Prefabs/Default Mobile UI.prefab [hunk 2]
```diff
@@ -1782,6 +1784,136 @@ MonoBehaviour:
...
+  m_CorrespondingSourceObject: {fileID: 0}
+  m_PrefabInstance: {fileID: 0}
+  m_PrefabAsset: {fileID: 0}
+  serializedVersion: 6
+  m_Component:
+  - component: {fileID: 2975881544335965903}
+  - component: {fileID: 731525659962201783}
+  - component: {fileID: 577653725205976457}
+  m_Layer: 0
+  m_Name: Asset State Text
```
Comment (mrbobbybobberson) [inline src/SpectatorView.Unity/Assets/SpectatorView/Prefabs/Default Mobile UI.prefab]: I added an extra "Asset State Text" text object above the other two state lines and moved them down to make room (this one can span to 2 lines!).

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpectatorView.cs [hunk 3]
```diff
@@ -320,9 +321,9 @@ private void SetupRecordingService()
                 recordingVisualPrefab = MobileRecordingSettings.Instance.OverrideMobileRecordingServicePrefab;
             }
 
-            if (MobileRecordingSettings.IsInitialized && 
-                MobileRecordingSettings.Instance.EnableMobileRecordingService &&
-                recordingVisualPrefab != null)
+            bool enableMobileRecordingService = (!MobileRecordingSettings.IsInitialized) || MobileRecordingSettings.Instance.EnableMobileRecordingService;
```
Comment (mrbobbybobberson) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpectatorView.cs]: Seems to me that if I don't have settings to explicitly disable recording, we should default to having recording (and the UX) on, right?  As it was formerly coded, if someone didn't add explicit settings, the mobile recording and status text wouldn't show up.

File: samples/Build2019Demo.Unity/Assets/Generated.StateSynchronization.AssetCaches/Resources/Android/spectatorview.manifest.bytes (hunk 1 above)
Comment (matthejo) [reply samples/Build2019Demo.Unity/Assets/Generated.StateSynchronization.AssetCaches/Resources/Android/spectatorview.manifest.bytes]: I think last time I committed, I was testing something out with loading a shader directly and had included it explicitly in the bundle through the Unity UI. It should not be there, yours is correct.

---
In reply to: [343311829](https://github.com/microsoft/MixedReality-SpectatorView/pull/270#discussion_r343311829) [](ancestors = 343311829)

File: src/SpectatorView.Unity/Assets/SpectatorView/Prefabs/Default Mobile UI.prefab [hunk 4]
```diff
@@ -755,7 +756,7 @@ MonoBehaviour:
   m_UiScaleMode: 1
   m_ReferencePixelsPerUnit: 100
   m_ScaleFactor: 1
-  m_ReferenceResolution: {x: 411, y: 731}
+  m_ReferenceResolution: {x: 663.9, y: 731}
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Prefabs/Default Mobile UI.prefab]: >663 [](start = 29, length = 3)

can we keep this at 411, 411x731 was provided by our designers.

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetBundleVersion.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetBundleVersion.cs]: nit: there's a new copyright header:

// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetBundleVersion.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+using UnityEngine;
+
+namespace Microsoft.MixedReality.SpectatorView
+{
+    public class AssetBundleVersion : ScriptableObject
+    {
+        public string Identity;
+        public string DisplayName;
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetBundleVersion.cs]: nit: public classes and properties should have summary descriptions.

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/UI/AssetStateVisual.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace Microsoft.MixedReality.SpectatorView
+{
+    public class AssetStateVisual : MonoBehaviour
+    {
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/UI/AssetStateVisual.cs]: do we expect folks to extend/use this class or could we make it internal?

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs [hunk 8]
```diff
@@ -102,26 +132,32 @@ protected override void OnConnected(SocketEndpoint endpoint)
             }
 
             hologramSynchronizer.Reset(endpoint);
-            ResetAssetCaches();
+            // TODO: should this be called here?!?!?!  ResetAssetCaches();
+
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs]: Was there a takeaway on whether or not this was needed?

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs [hunk 9]
```diff
@@ -256,6 +438,132 @@ private void ResetAssetCaches()
...
+        private class AssetBundleReceive
+        {
+            public byte[] Data;
+            public int NextDataToReceiveIndex;
+        }
+    }
+
+    public enum AssetStateStatus
+    {
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs]: nit: may be worth making this an internal/private enum declaration. Otherwise we need summaries for each value

File: src/SpectatorView.Unity/Assets/SpectatorView.Editor/Scripts/StateSynchronizationMenuItems.cs [hunk 10]
```diff
@@ -87,53 +87,93 @@ public static void ClearAllAssetCaches()
...
+            {
+                AssetBundlePlatform = AssetBundlePlatform.WSA,
+                BuildTargetGroup = BuildTargetGroup.WSA,
+                BuildTarget = BuildTarget.WSAPlayer,
+            });
 #endif
 
-            foreach (var directory in builtDirectories)
+            // TODO: save current build target and restore it at the end... perhaps build the current target bundle before switching (or maybe at the end).
+
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView.Editor/Scripts/StateSynchronizationMenuItems.cs]: we should definitely build the current target before switching, for most of these scoped apps its easy to switch players. For actual end user applications it can take 15-30 minutes just to switch players.

**Comment by chrisfromwork** [comment]
> CodeFlow for GitHub does not download executable files. If you trust this file and would like to run it, please download it yourself through GitHub.

what are these byte files for? are these asset caches?

---
Refers to: samples/Build2019Demo.Unity/Assets/Generated.StateSynchronization.AssetCaches/Resources/Android/spectatorview.bytes:1 in cc907f6. [](commit_id = cc907f6a432f3b9a8e8b35644d3bbb095df5dc42, deletion_comment = False)

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs [hunk 11]
```diff
@@ -20,9 +21,13 @@ public class StateSynchronizationObserver : NetworkManager<StateSynchronizationO
         public const string AssetBundleRequestInfoCommand = "RequestAssetBundleInfo";
         public const string AssetBundleReportInfoCommand = "ReportAssetBundleInfo";
         public const string AssetBundleRequestDownloadCommand = "RequestAssetBundleDownload";
-        public const string AssetBundleReportDownloadCommand = "ReportAssetBundleDownload";
+        public const string AssetBundleReportDownloadStartCommand = "ReportAssetBundleDownloadStart";
+        public const string AssetBundleReportDownloadDataCommand = "ReportAssetBundleDownloadData";
         public const string AssetLoadCompletedCommand = "AssetLoadCompleted";
 
+        public const string AssetBundleName = "spectatorview";
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs]: >spectatorview [](start = 47, length = 13)

it seems like we should include some sort of indication that this is an asset bundle. maybe "SpectatorView.AssetBundle" or something?

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs [hunk 12]
```diff
@@ -20,9 +21,13 @@ public class StateSynchronizationObserver : NetworkManager<StateSynchronizationO
...
         public const string AssetBundleReportInfoCommand = "ReportAssetBundleInfo";
         public const string AssetBundleRequestDownloadCommand = "RequestAssetBundleDownload";
-        public const string AssetBundleReportDownloadCommand = "ReportAssetBundleDownload";
+        public const string AssetBundleReportDownloadStartCommand = "ReportAssetBundleDownloadStart";
+        public const string AssetBundleReportDownloadDataCommand = "ReportAssetBundleDownloadData";
         public const string AssetLoadCompletedCommand = "AssetLoadCompleted";
 
+        public const string AssetBundleName = "spectatorview";
+        public const int AssetBundleReportDownloadDataMaxByteCount = 256 * 1024;
+
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs]: can we add a comment on what this is for?

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpectatorView.cs (hunk 3 above)
Comment (chrisfromwork) [reply src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpectatorView.cs]: This is a good point. We may want to clean up these functions better now that we have overloaded the mobile recording visual to contain items other than recording content.

---
In reply to: [343313429](https://github.com/microsoft/MixedReality-SpectatorView/pull/270#discussion_r343313429) [](ancestors = 343313429)

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs [hunk 13]
```diff
@@ -15,6 +15,9 @@ namespace Microsoft.MixedReality.SpectatorView
 {
     internal abstract class AssetCache : ScriptableObject
     {
+        public static int AssetCacheCount { get; private set; }
+        public static event Action<int> AssetCacheCountChanged;
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs]: >AssetCacheCountChanged [](start = 40, length = 22)

why was this needed?

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs [hunk 14]
```diff
@@ -154,32 +190,178 @@ private void HandleAssetBundleInfoCommand(SocketEndpoint endpoint, string comman
...
+                        Status = AssetStateStatus.AssetBundleLoaded,
+                        AssetBundleDisplayName = currentAssetBundleDisplayName,
+                    };
+
+                    SendAssetsLoaded(endpoint);
+                }
+                else
+                {
+                    DebugLog($"Requesting asset bundle download for {AssetBundleVersion.Format(assetBundleIdentity, assetBundleDisplayName)}...");
+                    Debug.Assert(AssetState.Status == AssetStateStatus.RequestingAssetBundle, this);
```
Comment (chrisfromwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/StateSynchronizationObserver.cs]: >AssetState.Status == AssetStateStatus.RequestingAssetBundle [](start = 33, length = 59)

where does this get set to RequestAssetBundle?
